File: fuspy/pages/fusionww/registration_page.py
"""
FusionWW Registration Page Object Model
"""
from playwright.sync_api import Page, Locator


class RegistrationPage:
    """FusionWW registration page object"""
    
    def __init__(self, page: Page):
        self.page = page
        self.account_button: Locator = page.locator("#account-button")
        self.register_button: Locator = page.get_by_role("button", name="Register")
        self.modal_title: Locator = page.locator(".Modal_titleWrapper__VtBVY > div")
        self.register_now_button: Locator = page.get_by_role("button", name="Register")
        self.first_name_input: Locator = page.get_by_role("textbox", name="First Name")
        self.last_name_input: Locator = page.get_by_role("textbox", name="Last Name")
        self.email_input: Locator = page.get_by_role("textbox", name="Email Address")
        self.geographical_region_button: Locator = page.get_by_role("button", name="Geographical Region")
        self.americas_option: Locator = page.get_by_role("option", name="Americas")
        self.password_input: Locator = page.get_by_role("textbox", name="Password", exact=True)
        self.confirm_password_input: Locator = page.get_by_role("textbox", name="Confirm Password")
        self.close_modal_button: Locator = page.get_by_role("button", name="Close modal")
        self.modal_container: Locator = page.locator('[role="dialog"], .modal, [class*="Modal"]').first

    def goto(self) -> None:
        """Navigate to FusionWW homepage"""
        self.page.goto("https://www.fusionww.com/")

    def dismiss_interfering_elements(self) -> None:
        """Remove any cookie banners or overlays that might interfere"""
        try:
            self.page.evaluate("""
                () => {
                    const interferingSelectors = [
                        "#hs-banner-parent",
                        ".hs-cookie-notification",
                        ".cookie-banner",
                        "#cookie-consent",
                        '[id*="cookie"]',
                        '[class*="cookie"]',
                        '[id*="banner"]',
                        '[class*="banner"]',
                    ];

                    interferingSelectors.forEach((selector) => {
                        const elements = document.querySelectorAll(selector);
                        elements.forEach((element) => {
                            if (element instanceof HTMLElement) {
                                element.style.display = "none";
                                element.style.visibility = "hidden";
                                element.style.pointerEvents = "none";
                                element.remove();
                            }
                        });
                    });
                }
            """)
        except Exception as e:
            print(f"Error dismissing interfering elements: {e}")

    def accept_cookies_if_present(self) -> None:
        """Accept cookies if present with comprehensive fallback strategies"""
        try:
            # Wait for page to load
            self.page.wait_for_load_state("domcontentloaded")
            self.page.wait_for_timeout(3000)

            # Try multiple cookie banner selectors
            cookie_selectors = [
                'button[data-testid="accept-all-cookies"]',
                'button:has-text("Accept all")',
                'button:has-text("Accept All")',
                'button:has-text("Accept")',
                '#hs-eu-cookie-confirmation-buttons-area button:has-text("Accept")',
                '.hs-cookie-notification-buttons button:has-text("Accept")',
                '#hs-eu-cookie-confirmation-buttons-area button[type="button"]',
                'button[data-hs-cookie-consent="accept-all"]',
                '[data-testid="cookie-banner"] button:has-text("Accept")',
                '.cookie-banner button:has-text("Accept")',
                '#cookie-consent button:has-text("Accept")',
            ]

            cookies_accepted = False

            for selector in cookie_selectors:
                try:
                    cookie_button = self.page.locator(selector).first
                    if cookie_button.is_visible(timeout=2000):
                        print(f"Found cookie button with selector: {selector}")
                        cookie_button.click()
                        self.page.wait_for_timeout(2000)

                        # Verify cookie banner is gone
                        cookie_banner = self.page.locator(
                            "#hs-banner-parent, .hs-cookie-notification, .cookie-banner, #cookie-consent"
                        )
                        if cookie_banner.is_visible(timeout=1000):
                            print("Cookie banner still visible, trying to dismiss again")
                            self.page.evaluate("""
                                () => {
                                    const banners = document.querySelectorAll(
                                        "#hs-banner-parent, .hs-cookie-notification, .cookie-banner, #cookie-consent"
                                    );
                                    banners.forEach((banner) => {
                                        if (banner instanceof HTMLElement) {
                                            banner.style.display = "none";
                                            banner.remove();
                                        }
                                    });
                                }
                            """)

                        cookies_accepted = True
                        print("Cookies accepted successfully")
                        break
                except Exception as e:
                    print(f"Selector {selector} failed: {e}")
                    continue

            if not cookies_accepted:
                print("No cookie banner found or could not be accepted")
                self.page.evaluate("""
                    () => {
                        const banners = document.querySelectorAll(
                            "#hs-banner-parent, .hs-cookie-notification, .cookie-banner, #cookie-consent"
                        );
                        banners.forEach((banner) => {
                            if (banner instanceof HTMLElement) {
                                banner.style.display = "none";
                                banner.remove();
                            }
                        });
                    }
                """)
        except Exception as e:
            print(f"Error in accept_cookies_if_present: {e}")

    def open_registration_modal(self) -> None:
        """Open the registration modal"""
        # Dismiss any interfering elements first
        self.dismiss_interfering_elements()

        # Wait for account button to be visible and clickable
        self.account_button.wait_for(state="visible", timeout=15000)
        self.account_button.click()

        # Wait for register button to be visible and clickable
        self.register_button.wait_for(state="visible", timeout=15000)
        self.register_button.click()

        # Wait for modal to be fully loaded
        self.page.wait_for_timeout(2000)

        # Wait for modal container to be visible
        self.modal_container.wait_for(state="visible", timeout=10000)

    def click_modal_title(self) -> None:
        """Click the modal title"""
        self.modal_title.click()

    def click_register_now(self) -> None:
        """Click the Register Now button"""
        # Dismiss any interfering elements first
        self.dismiss_interfering_elements()

        # Wait for register button to be visible and enabled
        self.register_now_button.wait_for(state="visible", timeout=15000)

        # Wait for the button to be enabled
        self.register_now_button.wait_for(state="attached")
        self.page.wait_for_function(
            "() => { const button = document.querySelector('button[disabled]'); return !button || !button.hasAttribute('disabled'); }",
            timeout=15000
        )

        self.register_now_button.click()

        # Wait for form to be fully loaded
        self.page.wait_for_timeout(2000)

        # Wait for form inputs to be visible
        self.first_name_input.wait_for(state="visible", timeout=10000)

    def fill_first_name(self, first_name: str) -> None:
        """Fill the first name field"""
        self.first_name_input.wait_for(state="visible")
        self.first_name_input.click()
        self.first_name_input.fill(first_name)

    def fill_last_name(self, last_name: str) -> None:
        """Fill the last name field"""
        self.last_name_input.wait_for(state="visible")
        self.last_name_input.click()
        self.last_name_input.fill(last_name)

    def fill_email(self, email: str) -> None:
        """Fill the email field"""
        self.email_input.wait_for(state="visible")
        self.email_input.click()
        self.email_input.fill(email)

    def select_geographical_region(self, region: str = "Americas") -> None:
        """Select geographical region"""
        self.geographical_region_button.wait_for(state="visible")
        self.geographical_region_button.click()

        # Wait for dropdown to open and select the specified region
        region_option = self.page.get_by_role("option", name=region)
        region_option.wait_for(state="visible")
        region_option.click()

    def fill_password(self, password: str) -> None:
        """Fill the password field"""
        self.password_input.wait_for(state="visible")
        self.password_input.click()
        self.password_input.fill(password)

    def fill_confirm_password(self, password: str) -> None:
        """Fill the confirm password field"""
        self.confirm_password_input.wait_for(state="visible")
        self.confirm_password_input.click()
        self.confirm_password_input.fill(password)

    def close_modal(self) -> None:
        """Close the modal"""
        self.close_modal_button.click()

    def close_modal_twice(self) -> None:
        """Close the modal with multiple fallback strategies"""
        try:
            # Try multiple close button selectors
            close_modal_selectors = [
                'button[aria-label="Close modal"]',
                'button:has-text("Close modal")',
                ".Modal_icon__v9JkP",
                '[data-testid="close-modal"]',
                'button[class*="close"]',
                'button[class*="Close"]',
                'button[aria-label="close"]',
                'button[title="Close"]',
                'button[title="close"]',
                '[role="button"][aria-label*="close" i]',
                '[role="button"][aria-label*="Close" i]',
                'button:has([class*="close"])',
                'button:has([class*="Close"])',
                "button:has(svg)",
                ".Modal_closeButton__*",
                ".close-button",
                ".closeButton",
                ".modal-close",
                ".modalClose",
            ]

            modal_closed = False

            # First attempt - try all selectors
            for selector in close_modal_selectors:
                try:
                    close_button = self.page.locator(selector).first
                    if close_button.is_visible(timeout=1000):
                        close_button.click()
                        print(f"Modal closed using selector: {selector}")
                        modal_closed = True
                        break
                except:
                    continue

            # Wait for modal to start closing
            self.page.wait_for_timeout(1000)

            # Second attempt - try again with different selectors
            if not modal_closed:
                second_attempt_selectors = [
                    'button[aria-label="Close modal"]',
                    'button:has-text("Close modal")',
                    ".Modal_icon__v9JkP",
                    'button[class*="close"]',
                    'button[class*="Close"]',
                ]

                for selector in second_attempt_selectors:
                    try:
                        close_button = self.page.locator(selector).first
                        if close_button.is_visible(timeout=1000):
                            close_button.click()
                            print(f"Modal closed on second attempt using selector: {selector}")
                            modal_closed = True
                            break
                    except:
                        continue

            # If still not closed, try pressing Escape key
            if not modal_closed:
                print("Trying Escape key to close modal")
                self.page.keyboard.press("Escape")
                self.page.wait_for_timeout(1000)

            # Final attempt - try clicking outside the modal
            if not modal_closed:
                print("Trying to click outside modal to close it")
                self.page.click("body", position={"x": 10, "y": 10})
                self.page.wait_for_timeout(1000)
        except Exception as e:
            print(f"Error in close_modal_twice: {e}")
            # Try pressing Escape key as fallback
            self.page.keyboard.press("Escape")

    def wait_for_modal_to_close(self) -> None:
        """Wait for modal to close"""
        try:
            # Wait for modal container to disappear
            self.modal_container.wait_for(state="detached", timeout=10000)
        except Exception as e:
            print(f"Modal container not found or already closed: {e}")

File: lib/pages/fusionww/registration.page.ts
import { type Locator, type Page } from "@playwright/test";

export class RegistrationPage {
  readonly page: Page;
  readonly accountButton: Locator;
  readonly registerButton: Locator;
  readonly modalTitle: Locator;
  readonly registerNowButton: Locator;
  readonly firstNameInput: Locator;
  readonly lastNameInput: Locator;
  readonly emailInput: Locator;
  readonly phoneInput: Locator;
  readonly geographicalRegionButton: Locator;
  readonly americasOption: Locator;
  readonly passwordInput: Locator;
  readonly confirmPasswordInput: Locator;
  readonly closeModalButton: Locator;
  readonly modalContainer: Locator;
  readonly modalContent: Locator;
  readonly modalErrorMessage: Locator;
  readonly modalErrorPasswordMessage: Locator;
  readonly modalErrorConfirmNameMessage: Locator;

  constructor(page: Page) {
    this.page = page;
    this.accountButton = page.locator("#account-button");
    this.registerButton = page.getByRole("button", { name: "Register" });
    this.modalTitle = page.locator(".Modal_titleWrapper__VtBVY > div");
    this.registerNowButton = page.getByRole("button", { name: "Register" });
    this.firstNameInput = page.getByRole("textbox", { name: "First Name" });
    this.lastNameInput = page.getByRole("textbox", { name: "Last Name" });
    this.emailInput = page.getByRole("textbox", { name: "Email Address" });
    this.phoneInput = page.getByRole("textbox", { name: /Phone/i });
    this.modalErrorPasswordMessage = page.getByText(
      "Minimum 8 characters, One lowercase letter, One uppercase letter, One number, One special character"
    );
    this.modalErrorConfirmNameMessage = page.getByText(
      "Invalid characters"
    );
    this.modalContent = page.getByText("Please provide a valid email address");
    this.modalErrorMessage = page.getByText(
      "Minimum 8 characters, One lowercase letter, One uppercase letter, One number, One special character"
    ); // Adjust selector as needed
    this.geographicalRegionButton = page.getByRole("button", {
      name: "Geographical Region",
    });
    this.americasOption = page.getByRole("option", { name: "Americas" });
    this.passwordInput = page.getByRole("textbox", {
      name: "Password",
      exact: true,
    });
    this.confirmPasswordInput = page.getByRole("textbox", {
      name: "Confirm Password",
    });
    this.closeModalButton = page.getByRole("button", { name: "Close modal" });
    this.modalContainer = page
      .locator('[role="dialog"], .modal, [class*="Modal"]')
      .first();
  }

  async goto() {
    await this.page.goto("https://www.fusionww.com/");
  }

  async dismissInterferingElements() {
    try {
      // Remove any cookie banners or overlays that might interfere
      await this.page.evaluate(() => {
        const interferingSelectors = [
          "#hs-banner-parent",
          ".hs-cookie-notification",
          ".cookie-banner",
          "#cookie-consent",
          '[id*="cookie"]',
          '[class*="cookie"]',
          '[id*="banner"]',
          '[class*="banner"]',
        ];

        interferingSelectors.forEach((selector) => {
          const elements = document.querySelectorAll(selector);
          elements.forEach((element) => {
            if (element instanceof HTMLElement) {
              element.style.display = "none";
              element.style.visibility = "hidden";
              element.style.pointerEvents = "none";
              element.remove();
            }
          });
        });
      });
    } catch (error) {
      console.log(`Error dismissing interfering elements: ${error.message}`);
    }
  }

  async acceptCookiesIfPresent() {
    try {
      // Wait for page to load with a reasonable timeout
      await this.page.waitForLoadState("domcontentloaded");

      // Wait a bit for any dynamic content to load
      await this.page.waitForTimeout(3000);

      // Try multiple cookie banner selectors with better error handling
      const cookieSelectors = [
        'button[data-testid="accept-all-cookies"]',
        'button:has-text("Accept all")',
        'button:has-text("Accept All")',
        'button:has-text("Accept")',
        '#hs-eu-cookie-confirmation-buttons-area button:has-text("Accept")',
        '.hs-cookie-notification-buttons button:has-text("Accept")',
        '#hs-eu-cookie-confirmation-buttons-area button[type="button"]',
        'button[data-hs-cookie-consent="accept-all"]',
        '[data-testid="cookie-banner"] button:has-text("Accept")',
        '.cookie-banner button:has-text("Accept")',
        '#cookie-consent button:has-text("Accept")',
      ];

      let cookiesAccepted = false;

      for (const selector of cookieSelectors) {
        try {
          const cookieButton = this.page.locator(selector).first();
          if (await cookieButton.isVisible({ timeout: 2000 })) {
            console.log(`Found cookie button with selector: ${selector}`);
            await cookieButton.click();
            // Wait for cookie banner to disappear completely
            await this.page.waitForTimeout(2000);

            // Verify cookie banner is gone by checking if it's not visible
            const cookieBanner = this.page.locator(
              "#hs-banner-parent, .hs-cookie-notification, .cookie-banner, #cookie-consent"
            );
            if (
              await cookieBanner.isVisible({ timeout: 1000 }).catch(() => false)
            ) {
              console.log(
                "Cookie banner still visible, trying to dismiss again"
              );
              // Try to dismiss any remaining cookie elements
              await this.page.evaluate(() => {
                const banners = document.querySelectorAll(
                  "#hs-banner-parent, .hs-cookie-notification, .cookie-banner, #cookie-consent"
                );
                banners.forEach((banner) => {
                  if (banner instanceof HTMLElement) {
                    banner.style.display = "none";
                    banner.remove();
                  }
                });
              });
            }

            cookiesAccepted = true;
            console.log("Cookies accepted successfully");
            break;
          }
        } catch (error) {
          // Continue to next selector if this one fails
          console.log(`Selector ${selector} failed: ${error.message}`);
          continue;
        }
      }

      if (!cookiesAccepted) {
        console.log("No cookie banner found or could not be accepted");
        // Try to remove any cookie banners that might be interfering
        await this.page.evaluate(() => {
          const banners = document.querySelectorAll(
            "#hs-banner-parent, .hs-cookie-notification, .cookie-banner, #cookie-consent"
          );
          banners.forEach((banner) => {
            if (banner instanceof HTMLElement) {
              banner.style.display = "none";
              banner.remove();
            }
          });
        });
      }
    } catch (error) {
      console.log(`Error in acceptCookiesIfPresent: ${error.message}`);
      // Don't fail the test if cookies can't be accepted
    }
  }

  async openRegistrationModal() {
    // Dismiss any interfering elements first
    await this.dismissInterferingElements();

    // Wait for account button to be visible and clickable
    await this.accountButton.waitFor({ state: "visible", timeout: 15000 });
    await this.accountButton.click();

    // Try multiple register button selectors
    const registerButtonSelectors = [
      this.registerButton,
      this.page.getByRole("button", { name: "Register" }),
      this.page.getByRole("button", { name: "Sign Up" }),
      this.page.getByRole("button", { name: "Create Account" }),
      this.page.locator('button:has-text("Register")'),
      this.page.locator('button:has-text("Sign Up")'),
      this.page.locator('button:has-text("Create Account")'),
      this.page.locator('[data-testid*="register"]'),
      this.page.locator('[data-testid*="signup"]'),
    ];

    let registerButtonFound = false;
    for (const selector of registerButtonSelectors) {
      try {
        if (await selector.isVisible({ timeout: 3000 })) {
          await selector.click();
          registerButtonFound = true;
          break;
        }
      } catch (error) {
        continue;
      }
    }

    if (!registerButtonFound) {
      console.log("Register button not found, trying alternative approach");
      // Try clicking on any button that might open the registration modal
      const anyButton = this.page.locator("button").first();
      if (await anyButton.isVisible({ timeout: 2000 })) {
        await anyButton.click();
      }
    }

    // Wait for modal to be fully loaded
    await this.page.waitForTimeout(2000);

    // Wait for modal container to be visible
    await this.modalContainer.waitFor({ state: "visible", timeout: 10000 });
  }

  async clickModalTitle() {
    await this.modalTitle.click();
  }

  async clickRegisterNow() {
    // Dismiss any interfering elements first
    await this.dismissInterferingElements();

    // Wait for register button to be visible and enabled
    await this.registerNowButton.waitFor({ state: "visible", timeout: 15000 });

    // Wait for the button to be enabled (not disabled)
    await this.registerNowButton.waitFor({ state: "attached" });
    await this.page.waitForFunction(
      () => {
        const button = document.querySelector("button[disabled]");
        return !button || !button.hasAttribute("disabled");
      },
      { timeout: 15000 }
    );

    await this.registerNowButton.click();

    // Wait for form to be fully loaded
    await this.page.waitForTimeout(2000);

    // Wait for form inputs to be visible
    await this.firstNameInput.waitFor({ state: "visible", timeout: 10000 });
  }

  async fillFirstName(firstName: string) {
    await this.firstNameInput.waitFor({ state: "visible" });
    await this.firstNameInput.click();
    await this.firstNameInput.fill(firstName);
  }

  async fillLastName(lastName: string) {
    await this.lastNameInput.waitFor({ state: "visible" });
    await this.lastNameInput.click();
    await this.lastNameInput.fill(lastName);
  }

  async fillEmail(email: string) {
    await this.emailInput.waitFor({ state: "visible" });
    await this.emailInput.click();
    await this.emailInput.fill(email);
  }

  async fillPhoneNumber(phone: string) {
    // Try multiple phone field selectors since the exact field might not exist
    const phoneSelectors = [
      this.phoneInput,
      this.page.getByRole("textbox", { name: "Phone" }),
      this.page.getByRole("textbox", { name: "Phone Number" }),
      this.page.locator('input[type="tel"]'),
      this.page.locator('input[name*="phone"]'),
      this.page.locator('input[id*="phone"]'),
      this.page.locator('input[placeholder*="phone" i]'),
    ];

    let phoneFieldFound = false;
    for (const selector of phoneSelectors) {
      try {
        if (await selector.isVisible({ timeout: 2000 })) {
          await selector.click();
          await selector.fill(phone);
          phoneFieldFound = true;
          break;
        }
      } catch (error) {
        continue;
      }
    }

    if (!phoneFieldFound) {
      console.log("Phone field not found, skipping phone number input");
      // Don't fail the test if phone field doesn't exist
    }
  }

  async selectGeographicalRegion(region: string = "Americas") {
    await this.geographicalRegionButton.waitFor({ state: "visible" });
    await this.geographicalRegionButton.click();

    // Wait for dropdown to open and select the specified region
    const regionOption = this.page.getByRole("option", { name: region });
    await regionOption.waitFor({ state: "visible" });
    await regionOption.click();
  }

  async fillPassword(password: string) {
    await this.passwordInput.waitFor({ state: "visible" });
    await this.passwordInput.click();
    await this.passwordInput.fill(password);
  }

  async fillConfirmPassword(password: string) {
    await this.confirmPasswordInput.waitFor({ state: "visible" });
    await this.confirmPasswordInput.click();
    await this.confirmPasswordInput.fill(password);
  }

  async closeModal() {
    await this.closeModalButton.click();
  }

  async closeModalTwice() {
    try {
      // Try multiple close button selectors
      const closeModalSelectors = [
        'button[aria-label="Close modal"]',
        'button:has-text("Close modal")',
        ".Modal_icon__v9JkP",
        '[data-testid="close-modal"]',
        'button[class*="close"]',
        'button[class*="Close"]',
        'button[aria-label="close"]',
        'button[title="Close"]',
        'button[title="close"]',
        '[role="button"][aria-label*="close" i]',
        '[role="button"][aria-label*="Close" i]',
        'button:has([class*="close"])',
        'button:has([class*="Close"])',
        "button:has(svg)",
        ".Modal_closeButton__*",
        ".close-button",
        ".closeButton",
        ".modal-close",
        ".modalClose",
      ];

      let modalClosed = false;

      // First attempt - try all selectors
      for (const selector of closeModalSelectors) {
        try {
          const closeButton = this.page.locator(selector).first();
          if (await closeButton.isVisible({ timeout: 1000 })) {
            await closeButton.click();
            console.log(`Modal closed using selector: ${selector}`);
            modalClosed = true;
            break;
          }
        } catch (error) {
          continue;
        }
      }

      // Wait for modal to start closing
      await this.page.waitForTimeout(1000);

      // Second attempt - try again with different selectors
      if (!modalClosed) {
        const secondAttemptSelectors = [
          'button[aria-label="Close modal"]',
          'button:has-text("Close modal")',
          ".Modal_icon__v9JkP",
          'button[class*="close"]',
          'button[class*="Close"]',
        ];

        for (const selector of secondAttemptSelectors) {
          try {
            const closeButton = this.page.locator(selector).first();
            if (await closeButton.isVisible({ timeout: 1000 })) {
              await closeButton.click();
              console.log(
                `Modal closed on second attempt using selector: ${selector}`
              );
              modalClosed = true;
              break;
            }
          } catch (error) {
            continue;
          }
        }
      }

      // If still not closed, try pressing Escape key
      if (!modalClosed) {
        console.log("Trying Escape key to close modal");
        await this.page.keyboard.press("Escape");
        await this.page.waitForTimeout(1000);
      }

      // Final attempt - try clicking outside the modal
      if (!modalClosed) {
        console.log("Trying to click outside modal to close it");
        await this.page.click("body", { position: { x: 10, y: 10 } });
        await this.page.waitForTimeout(1000);
      }
    } catch (error) {
      console.log(`Error in closeModalTwice: ${error.message}`);
      // Try pressing Escape key as fallback
      await this.page.keyboard.press("Escape");
    }
  }

  async waitForModalToClose() {
    try {
      // Wait for modal container to disappear
      await this.modalContainer.waitFor({ state: "detached", timeout: 10000 });
    } catch (error) {
      console.log(
        `Modal container not found or already closed: ${error.message}`
      );
    }
  }
}
